### PR TITLE
[Core] Create node_id at Python Node __init__ and pass to raylet

### PR DESCRIPTION
**Pull request by ruisearch42**
## Why are these changes needed?
After ray processes are started during ray.init() or ray.up(), the script/worker should connects to the raylet it just started.
However currently the raylet is looked up based on IP address, and it might be possible that there exists another raylet with the same IP address, and as a result, a different raylet is connected.

This PR addresses this issue by creating node_id at Node `__init__` time, and pass it down and set it at raylet. We also look up raylet with node_id which is unique.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/main.cc [hunk 1]
```diff
@@ -365,11 +368,17 @@ int main(int argc, char *argv[]) {
             {ray::stats::SessionNameKey, session_name}};
         ray::stats::Init(global_tags, metrics_agent_port, WorkerID::Nil());
 
-        ray::NodeID raylet_node_id{
-            (!RayConfig::instance().OVERRIDE_NODE_ID_FOR_TESTING().empty())
-                ? ray::NodeID::FromHex(
-                      RayConfig::instance().OVERRIDE_NODE_ID_FOR_TESTING())
-                : ray::NodeID::FromRandom()};
+        ray::NodeID raylet_node_id;
+        if (!RayConfig::instance().OVERRIDE_NODE_ID_FOR_TESTING().empty()) {
```
Comment (jjyao) [inline src/ray/raylet/main.cc:372]: we should handle this `OVERRIDE_NODE_ID_FOR_TESTING` inside `node.py` I think

File: .buildkite/ml.rayci.yml [hunk 2]
```diff
@@ -50,128 +50,6 @@ steps:
       - "3.10"
 
   # tests
-  - label: ":train: ml: train tests"
```
Comment (rynewang) [inline .buildkite/ml.rayci.yml:53]: Accidental remove?

Comment (ruisearch42) [reply .buildkite/ml.rayci.yml:53]: No this is intentional, to temporarily limit tests in CI for debugging purpose. I will revert this after debugging is done.

File: python/ray/_private/node.py [hunk 3]
```diff
@@ -274,6 +274,24 @@ def __init__(
...
                 self._ray_params.raylet_socket_name, default_prefix="raylet"
             )
+            if (
+                self._ray_params.env_vars is not None
+                and "RAY_OVERRIDE_NODE_ID_FOR_TESTING" in self._ray_params.env_vars
+            ):
+                node_id = self._ray_params.env_vars["RAY_OVERRIDE_NODE_ID_FOR_TESTING"]
+                logger.debug(
+                    f"Setting node ID to {node_id} "
+                    + "based on ray_params.env_vars override"
```
Comment (jjyao) [inline python/ray/_private/node.py:284]: No need the + sign

**Review by jjyao — COMMENTED** [review]
Could you update the pr description to say about why we make this change?

File: python/ray/_private/services.py [hunk 4]
```diff
@@ -1533,6 +1544,7 @@ def start_raylet(
     Args:
         redis_address: The address of the primary Redis server.
         gcs_address: The address of GCS server.
+        node_id: The ID of this node.
```
Comment (jjyao) [inline python/ray/_private/services.py:1547]:
```suggestion
        node_id: The hex ID of this node.
```

File: python/ray/_private/state.py [hunk 5]
```diff
@@ -800,6 +800,11 @@ def get_node_to_connect_for_driver(self, node_ip_address):
             node_ip_address
         )
 
+    def get_node(self, node_id):
```
Comment (jjyao) [inline python/ray/_private/state.py:803]:
```suggestion
    def get_node(self, node_id: str):
```

File: src/ray/gcs/gcs_client/global_state_accessor.cc [hunk 6]
```diff
@@ -374,33 +374,87 @@ std::string GlobalStateAccessor::GetSystemConfig() {
   return future.get();
 }
 
-ray::Status GlobalStateAccessor::GetNodeToConnectForDriver(
-    const std::string &node_ip_address, std::string *node_to_connect) {
+ray::Status GlobalStateAccessor::GetAliveNodes(std::vector<rpc::GcsNodeInfo> &nodes) {
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/global_state_accessor.cc:377]:
```suggestion
ray::Status GlobalStateAccessor::GetAliveNodes(std::vector<rpc::GcsNodeInfo> *nodes) {
```

We use pointer for output

File: src/ray/gcs/gcs_client/global_state_accessor.cc [hunk 7]
```diff
@@ -374,33 +374,87 @@ std::string GlobalStateAccessor::GetSystemConfig() {
...
 
-    // Deal with alive nodes only
+    if (nodes.empty()) {
+      status = Status::NotFound("GCS has started but no raylets have registered yet.");
+    } else {
+      std::pair<std::string, int> gcs_address;
+      {
+        absl::WriterMutexLock lock(&mutex_);
+        gcs_address = gcs_client_->GetGcsServerAddress();
+      }
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/global_state_accessor.cc:420]: This is not used

**Review by jjyao — APPROVED** [review]

File: src/ray/gcs/gcs_client/global_state_accessor.cc (hunk 6 above)
Comment (ruisearch42) [reply src/ray/gcs/gcs_client/global_state_accessor.cc:377]: Discussed offline, this follows C++ style guide.

**Pull request merged**